grid-proxy/tools/db/generate.go

```diff
@@ -397,37 +609,117 @@ func generateNodeGPUs(db *sql.DB) error {
 			contract:     i % 2,
 			id:           "0000:0e:00.0/1002/744c",
 		}
-		if _, err := db.Exec(insertQuery(&g)); err != nil {
-			panic(err)
+
+		GPUs = append(GPUs, objectToTupleString(g))
+	}
+
+	if err := insertTuples(db, node_gpu{}, GPUs); err != nil {
+		return err
+	}
+
+	fmt.Println("node GPUs generated")
+
+	return nil
+
+}
+
+func generateContracts(db *sql.DB) error {
+	billsStartID := 1
+	contractsStartID := 1
+
+	err, billsStartID, contractCount := generateRentContracts(db, billsStartID, contractsStartID)
```

Comment:
this is better from last version. 
i would go further with this and make the generateRentContract, generateNodeContracts, and generateNameContracts functions just create the string arrays of the contracts and the bill reports, and do the actual insert calls here in the generateContracts function.
but this is just nitpicking, feel free to ignore.

Reply:
what about generating all contracts in the functions and collecting reports from each one and inserting them once? that's the one which will make the performance better as inserting all bills at once.